### ISSUE TITLE
Template injection: don't flag on `matrix` expansions when the matrix is statically defined

### ISSUE THREAD

**Issue by woodruffw** [woodruffw/zizmor#22]
This will be 100% FP.

h/t @alex

**Comment by VorpalBlade** [comment]
I believe I have come across another instance of this issue, though I'm not sure so I'll post my example here:

```
    |
105 |     runs-on: ${{ matrix.os }}
    |     ------------------------- note: expression may expand into a self-hosted runner
```

where

```yaml
    strategy:
      matrix:
        include:
          - target: aarch64-unknown-linux-gnu
            os: ubuntu-latest
            rustflags: -Clink-arg=-Wl,--compress-debug-sections=zlib
          - target: aarch64-unknown-linux-musl
            os: ubuntu-latest
            rustflags: -Clink-arg=-Wl,--compress-debug-sections=zlib
          - target: armv7-unknown-linux-gnueabihf
            os: ubuntu-latest
            rustflags: -Clink-arg=-Wl,--compress-debug-sections=zlib
          - target: armv7-unknown-linux-musleabihf
            os: ubuntu-latest
            rustflags: -Clink-arg=-Wl,--compress-debug-sections=zlib
          - target: i686-unknown-linux-gnu
            os: ubuntu-latest
            rustflags: -Clink-arg=-Wl,--compress-debug-sections=zlib
          - target: i686-unknown-linux-musl
            os: ubuntu-latest
            rustflags: -Clink-arg=-Wl,--compress-debug-sections=zlib
          - target: x86_64-unknown-linux-gnu
            os: ubuntu-latest
            rustflags: -Clink-arg=-Wl,--compress-debug-sections=zlib
          - target: x86_64-unknown-linux-musl
            os: ubuntu-latest
            rustflags: -Clink-arg=-Wl,--compress-debug-sections=zlib
          - target: x86_64-pc-windows-msvc
            os: windows-latest
            rustflags:
          - target: x86_64-apple-darwin
            os: macos-latest
            rustflags:
          - target: aarch64-apple-darwin
            os: macos-latest
            rustflags:
```

All the OSes are hosted runners by github, so this is a false positive.

**Comment by woodruffw** [comment]
Thanks @VorpalBlade! That does indeed look like a lingering false positive -- would you be able to share your entire workflow so that I can fully reproduce the bug here?

**Comment by VorpalBlade** [comment]
Sure, that would be https://github.com/VorpalBlade/chezmoi_modify_manager/blob/main/.github/workflows/release.yml

**Comment by woodruffw** [comment]
Oh, I see what's happening here: this is in `self-hosted-runner`, not `template-injection`. The latter doesn't flag this even at `--pedantic`, but the former does because we don't currently attempt to walk the matrix at all in that case.

**Comment by Julian** [comment]
(I'm finding this issue because it flags my typical pattern for running nox, so I'm just curious to learn what the plan is here).

Does "static" here mean literally "fully static" as in a YAML array or something that someone has used? Or might it include what I do which is:

```yaml
      - id: noxenvs-matrix
        run: |
          echo >>$GITHUB_OUTPUT noxenvs=$(
            nox --list-sessions --json | jq '[.[].session]'
          )

[sometime later ...]

      matrix:
        noxenv: ${{ fromJson(needs.list.outputs.noxenvs) }}

```

where there's at least no user input involved in calculating the matrix. Obviously I know that trying to parse some arbitrary shell is untenable, so maybe this is impossible as unless the precise command is known not to secretly take input from somewhere there's no way of knowing anything about it, but just curious what to expect in the future. (And thanks for a great looking tool!)

(P.S. what's FP? I assume not functional programming. Future proof?)

**Comment by woodruffw** [comment]
> Does "static" here mean literally "fully static" as in a YAML array or something that someone has used? Or might it include what I do which is:

The former unfortunately -- "static" from `zizmor`'s vantage point means "no runtime dataflow into the matrix at all," since it's frequently impossible to analyze those inputs.

For the time being, the best workaround for that is probably to use an [ignore rule](https://woodruffw.github.io/zizmor/configuration/#rulesidignore).

> (P.S. what's FP? I assume not functional programming. Future proof?)

False positive :slightly_smiling_face:

**Comment by hugovk** [comment]
I'm getting "matrix.llvm may expand into attacker-controllable code" and "matrix.architecture may expand into attacker-controllable code" with zizmor 0.9.0 and 0.9.1 for:

https://github.com/python/cpython/blob/e2325c9db0650fc06d909eb2b5930c0573f24f71/.github/workflows/jit.yml

This doesn't happen with 0.8.0.

**Comment by woodruffw** [comment]
Thanks @hugovk! I'll look into that today.

**Comment by ubiratansoares** [comment]
Hi @hugovk, thanks for catching that!

It seems indeed a regression! 

If I understand everything correctly, the issue is that I designed the check for a fully static matrix by walk all possible expansions and checking whether any of them resolves to an expression, which seems to be the case [here](https://github.com/python/cpython/blob/e2325c9db0650fc06d909eb2b5930c0573f24f71/.github/workflows/jit.yml#L87).

However, in the case of template injections, I think the correct approach is actually resolving whether we expand to static values according to the context we are evaluating (e.g, `matrix.llvm` or `matrix.architecture` in your case). 

I have a tentative fix [here](https://github.com/woodruffw/zizmor/pull/298). I used your Workflow as a bug reproducer 🙂 

cc @woodruffw

**Comment by woodruffw** [comment]
Thanks both! I'm prepping a 0.9.2 release now; it should be live shortly.

**Comment by hugovk** [comment]
Thanks, confirmed this no longer shows up with 0.9.2 🚀

**Comment by woodruffw** [comment]
The bulk of the remaining cases are done here, so closing. We can track individual FP cases in new issues!
